web/settings/modulconfigbat.php

```diff
@@ -279,7 +280,23 @@ function visibility_solarwatt_ip2() {
 								Konfiguration im zugehörigen EVU Modul.
 							</div>
 						</div>
+						<div id="divspeicherrct2" class="hide">
+				<div class="card-header bg-secondary">
+					Speicher Hardware
+				</div>
+				<div class="card-body">
+					<div class="row">
+						<div class="col">
+						<pre><?php 
+						   $lines=[];
+							exec('/var/www/html/openWB/modules/speicher_rct2/info.sh', $lines);
+							echo implode('<br>',$lines); 
+						  ?></pre>
```

Comment:
Das `<pre>` sorgt bereits dafür, dass Zeilenumbrüche erhalten bleiben. Insofern ist das `<br>` unnötig bzw. wird nur gebraucht weil exec dummerweise einen Array erzeugt. Sinnvoll dagegen wäre es HTML zu escapen. Wenn im Output sonst zufälligerweise ein `<` vorkommt, passieren lustige Dinge. Besser wäre daher:
```php
echo htmlspecialchars(shell_exec('/var/www/html/openWB/modules/speicher_rct2/info.sh'));
```

Reply:
Ja ist ein Array of Lines.
Sehe in info.sh nach 
Da es keine User-EIngaben sind sollte html....  nicht nötig sein.

Reply:
> Ja ist ein Array of Lines.

Deswegen `shell_exec` in meinem Beispiel, dann ist es kein Array sondern ein String.

> Da es keine User-EIngaben sind sollte html.... nicht nötig sein.

Im Code steht zum Beispiel:
```python
print( "Battery Controler " + str(rct_lib.host) )
```
Ich gehe davon aus, dass der host auf der UI eingegeben wird und das entsprechend eine User-Eingabe ist.

Selbst wenn es das nicht ist. Willst du dich darauf immer verlassen? Was ist wenn jemand dein Script verändert und irgendwas hinzufügt? Was ist wenn dein Skript einen Fehler wirft wegen... was auch immer? Dann ist nicht garantiert, dass da unzulässige Zeichen drin sind. Und gerade in dem Fall will man diesen Statustext ordentlich lesen können.

`htmlspecialchars` ist einfach, billig und schafft alle diese Bedenken ganz leicht aus der Welt.